README
NAME
    Git::Hooks::PerlCritic - Perl Critic hooks for git

VERSION
    version 0.001000

DESCRIPTION
    Allows you to utilize Perl::Critic via git hooks
    <http://www.kernel.org/pub/software/scm/git/docs/githooks.html> using
    the Git::Hooks framework.

    First setup git-hooks.pl

    Then you should choose to use only one of the available hooks.

HOOKS
  pre-commit
            ln -s git-hooks.pl pre-commit
            git config --add githooks.pre-commit PerlCritic

    This hook will prevent a commit that doesn't pass Perl::Critic from
    being committed.

  prepare-commit-msg
            ln -s git-hooks.pl prepare-commit-msg
            git config --add githooks.preprare-commit-msg PerlCritic

    This hook will simply append commented out critic warnings to the commit
    message, so you may review them before committing. Note: requires
    Path::Class::File.

SEE ALSO
    Git::Hooks
    Perl::Critic

BUGS
    Please report any bugs or feature requests on the bugtracker website
    https://github.com/hostgator/git-hooks-perlcritic/issues or by email to
    development@hostgator.com.

    When submitting a bug or request, please include a test-file or a patch
    to an existing test-file that illustrates the bug or desired feature.

AUTHOR
    Caleb Cushing <xenoterracide@gmail.com>

COPYRIGHT AND LICENSE
    This software is Copyright (c) 2013 by HostGator.com
    <http://hostgator.com>.

    This is free software, licensed under:

      The Artistic License 2.0 (GPL Compatible)



Changes
Revision history for Git-Hooks-PerlCritic

0.001000  2013-01-03
	- initial release


lib/Git/Hooks/PerlCritic.pm
package Git::Hooks::PerlCritic;
use 5.010;
use strict;
use warnings;

our $VERSION = '0.001000'; # VERSION

use Carp;
use Module::Load 'load';
use Git::Hooks;

sub _changed {
	my $git = shift;

	my @changed
		= grep { /\.(p[lm]|t)$/xms }
		$git->command( qw/diff --cached --name-only --diff-filter=AM/ )
		;

	return \@changed;
}

sub _set_critic {
	load 'Perl::Critic';
	load 'Perl::Critic::Violation';
	load 'Perl::Critic::Utils';

	my $pc = Perl::Critic->new;
	my $verbosity = $pc->config->verbose;

	# set the format to be a comment
	my $fmt = Perl::Critic::Utils::verbosity_to_format( $verbosity );
	Perl::Critic::Violation::set_format( "# $fmt" );

	return $pc;
}

sub _check_violations {
	my $files = shift;

	my @violations;
	foreach my $file ( @$files ) {
		state $critic = _set_critic;

		@violations = $critic->critique( $file );
	}

	return \@violations;
}

PREPARE_COMMIT_MSG {
	my ( $git, $commit_msg_file ) = @_;

	my $changed    = _changed( $git );
	my $violations = _check_violations( $changed );

	if ( @$violations ) {
		my $pcf = 'Path::Class::File'; load $pcf;
		my $file     = $pcf->new( $commit_msg_file );
		my $contents = $file->slurp;

		# a space is being prepended, suspect internal join, remove it
		( $contents .= "@$violations" ) =~ s/^\ #//xmsg;

		$file->spew( $contents );
	}
};

PRE_COMMIT {
	my $git = shift;

	my $changed    = _changed( $git );
	my $violations = _check_violations( $changed );

	if ( @$violations ) {
		print @$violations;
		# . operator causes the array ref to give count, otherwise it would
		# stringify
		croak '# please fix ' . @$violations . ' perl critic errors before committing';
	}
};

1;

# ABSTRACT: Perl Critic hooks for git

__END__

=pod

=head1 NAME

Git::Hooks::PerlCritic - Perl Critic hooks for git

=head1 VERSION

version 0.001000

=head1 DESCRIPTION

Allows you to utilize L<Perl::Critic> via
L<git hooks|http://www.kernel.org/pub/software/scm/git/docs/githooks.html>
using the L<Git::Hooks> framework.

First setup L<git-hooks.pl|Git::Hooks/"USAGE">

Then you should choose to use only one of the available hooks.

=head1 HOOKS

=head2 pre-commit

	ln -s git-hooks.pl pre-commit
	git config --add githooks.pre-commit PerlCritic

This hook will prevent a commit that doesn't pass L<Perl::Critic> from being
committed.

=head2 prepare-commit-msg

	ln -s git-hooks.pl prepare-commit-msg
	git config --add githooks.preprare-commit-msg PerlCritic

This hook will simply append commented out critic warnings to the commit
message, so you may review them before committing. Note: requires
L<Path::Class::File>.

=head1 SEE ALSO

=over

=item L<Git::Hooks>

=item L<Perl::Critic>

=back

=head1 BUGS

Please report any bugs or feature requests on the bugtracker website
https://github.com/hostgator/git-hooks-perlcritic/issues or by email to
development@hostgator.com.

When submitting a bug or request, please include a test-file or a
patch to an existing test-file that illustrates the bug or desired
feature.

=head1 AUTHOR

Caleb Cushing <xenoterracide@gmail.com>

=head1 COPYRIGHT AND LICENSE

This software is Copyright (c) 2013 by L<HostGator.com|http://hostgator.com>.

This is free software, licensed under:

  The Artistic License 2.0 (GPL Compatible)

=cut


Makefile.PL

use strict;
use warnings;

use 5.010;

use ExtUtils::MakeMaker 6.30;



my %WriteMakefileArgs = (
  "ABSTRACT" => "Perl Critic hooks for git",
  "AUTHOR" => "Caleb Cushing <xenoterracide\@gmail.com>",
  "BUILD_REQUIRES" => {
    "File::Find" => 0,
    "File::Temp" => 0,
    "Test::More" => "0.88"
  },
  "CONFIGURE_REQUIRES" => {
    "ExtUtils::MakeMaker" => "6.30"
  },
  "DISTNAME" => "Git-Hooks-PerlCritic",
  "EXE_FILES" => [],
  "LICENSE" => "artistic_2",
  "NAME" => "Git::Hooks::PerlCritic",
  "PREREQ_PM" => {
    "Carp" => 0,
    "Git::Hooks" => 0,
    "Module::Load" => 0,
    "Perl::Critic" => 0,
    "Perl::Critic::Utils" => 0,
    "Perl::Critic::Violation" => 0,
    "strict" => 0,
    "warnings" => 0
  },
  "VERSION" => "0.001000",
  "test" => {
    "TESTS" => "t/*.t"
  }
);


unless ( eval { ExtUtils::MakeMaker->VERSION(6.56) } ) {
  my $br = delete $WriteMakefileArgs{BUILD_REQUIRES};
  my $pp = $WriteMakefileArgs{PREREQ_PM};
  for my $mod ( keys %$br ) {
    if ( exists $pp->{$mod} ) {
      $pp->{$mod} = $br->{$mod} if $br->{$mod} > $pp->{$mod};
    }
    else {
      $pp->{$mod} = $br->{$mod};
    }
  }
}

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };

WriteMakefile(%WriteMakefileArgs);



